src/server/pkg/obj/microsoft_client.go

```diff
@@ -116,71 +125,80 @@ type microsoftWriter struct {
 	blob       string
 	blobClient storage.BlobStorageClient
+	buf        *bytes.Buffer
+	nBlocks    int
+	eg         errgroup.Group
+	err        error
 }
 
 func newMicrosoftWriter(client *microsoftClient, name string) (*microsoftWriter, error) {
-	// create container
-	_, err := client.blobClient.CreateContainerIfNotExists(client.container, storage.ContainerAccessTypePrivate)
-	if err != nil {
+	if _, err := client.blobClient.CreateContainerIfNotExists(client.container, storage.ContainerAccessTypePrivate); err != nil {
 		return nil, err
 	}
-
-	// create blob
-	err = client.blobClient.CreateBlockBlob(client.container, name)
-	if err != nil {
+	if err := client.blobClient.CreateBlockBlob(client.container, name); err != nil {
 		return nil, err
 	}
-
 	return &microsoftWriter{
 		container:  client.container,
 		blob:       name,
 		blobClient: client.blobClient,
+		buf:        bytes.NewBuffer(bufPool.GetBuffer()[:0]),
 	}, nil
 }
 
 func (w *microsoftWriter) Write(b []byte) (int, error) {
-	blockList, err := w.blobClient.GetBlockList(w.container, w.blob, storage.BlockListTypeAll)
-	if err != nil {
-		return 0, err
+	if w.err != nil {
+		// w.err wasn't generated by putting this block but a previous block,
+		// we return it here though because PutBlock is called async so we
+		// can't return it from the correct call but we also don't want to wait
+		// until the end to handle errors.
+		return 0, w.err
 	}
-
-	blocksLen := len(blockList.CommittedBlocks)
-	amendList := []storage.Block{}
-	for _, v := range blockList.CommittedBlocks {
-		amendList = append(amendList, storage.Block{v.Name, storage.BlockStatusCommitted})
-	}
-
-	inputSourceReader := bytes.NewReader(b)
-	chunk := grpcutil.GetBuffer()
-	defer grpcutil.PutBuffer(chunk)
+	nBytes := 0
 	for {
-		n, err := inputSourceReader.Read(chunk)
-		if err == io.EOF {
+		if w.buf.Len()+len(b) >= maxBlockSize {
+			offset := maxBlockSize - w.buf.Len()
+			w.buf.Write(b[:offset])
+			w.writeBlock(w.buf.Bytes())
+			nBytes += offset
+			w.buf = bytes.NewBuffer(bufPool.GetBuffer()[:0])
+			b = b[offset:]
+		} else {
+			w.buf.Write(b)
+			nBytes += len(b)
 			break
 		}
-		if err != nil {
-			return 0, err
-		}
-
-		blockID := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%011d\n", blocksLen)))
-		data := chunk[:n]
-		err = w.blobClient.PutBlock(w.container, w.blob, blockID, data)
-		if err != nil {
-			return 0, fmt.Errorf("BlobStorageClient.PutBlock: %v", err)
-		}
-		// add current uncommitted block to temporary block list
-		amendList = append(amendList, storage.Block{blockID, storage.BlockStatusUncommitted})
-		blocksLen++
 	}
+	return nBytes, nil
+}
 
-	// update block list to blob committed block list.
-	err = w.blobClient.PutBlockList(w.container, w.blob, amendList)
-	if err != nil {
-		return 0, fmt.Errorf("BlobStorageClient.PutBlockList: %v", err)
-	}
-	return len(b), nil
+func blockID(n int) string {
+	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%011d\n", n)))
+}
```

Comment:
This block ID format looks strange. Could you add an explanation of what it's doing, and why azure needs it?

Reply:
I will attempt to, this format was compliantly copied over from the old version of the code that used this format for the block ids. I'll see if I can find some docs.